src/c_kzg_4844.c

```diff
@@ -2378,231 +2379,84 @@ typedef struct {
-        ret = C_KZG_BADARGS;
-        goto out;
-    }
-    if (partial_count == 0) {
-        ret = C_KZG_BADARGS;
-        goto out;
-    }
-
-    /*
-     * The degree of the output polynomial is the sum of the degrees of the
-     * input polynomials.
-     */
-    size_t out_degree = 0;
-    for (size_t i = 0; i < partial_count; i++) {
-        out_degree += partials[i].length - 1;
-    }
-    if (out_degree + 1 > len_out) {
-        ret = C_KZG_BADARGS;
-        goto out;
+    if (roots_len == 0) {
+        return C_KZG_BADARGS;
     }
 
-    /* Split scratch up into three equally sized working arrays */
-    fr_t *p_padded = scratch + len_out * 0;
-    fr_t *mul_eval_ps = scratch + len_out * 1;
-    fr_t *p_eval = scratch + len_out * 2;
+    /* Initialize with -root[0] */
+    blst_fr_cneg(&poly[0], &roots[0], true);
 
-    /*
-     * Do the last partial first: it is no longer than the others and the
-     * padding can remain in place for the rest.
-     */
-    ret = pad_p(
-        p_padded,
-        len_out,
-        partials[partial_count - 1].coeffs,
-        partials[partial_count - 1].length
-    );
-    if (ret != C_KZG_OK) goto out;
+    for (size_t i = 1; i < roots_len; i++) {
+        blst_fr_cneg(&neg_root, &roots[i], true);
 
-    ret = fft_fr(mul_eval_ps, p_padded, len_out, s);
-    if (ret != C_KZG_OK) goto out;
+        poly[i] = neg_root;
+        blst_fr_add(&poly[i], &poly[i], &poly[i - 1]);
 
-    for (size_t i = 0; i < partial_count - 1; i++) {
-        ret = pad_p(
-            p_padded, partials[i].length, partials[i].coeffs, partials[i].length
-        );
-        if (ret != C_KZG_OK) goto out;
-        ret = fft_fr(p_eval, p_padded, len_out, s);
-        if (ret != C_KZG_OK) goto out;
-        for (size_t j = 0; j < len_out; j++) {
-            blst_fr_mul(&mul_eval_ps[j], &mul_eval_ps[j], &p_eval[j]);
+        for (size_t j = i - 1; j > 0; j--) {
+            blst_fr_mul(&poly[j], &poly[j], &neg_root);
+            blst_fr_add(&poly[j], &poly[j], &poly[j - 1]);
         }
+        blst_fr_mul(&poly[0], &poly[0], &neg_root);
     }
 
-    ret = ifft_fr(out->coeffs, mul_eval_ps, len_out, s);
-    if (ret != C_KZG_OK) goto out;
-
-    out->length = out_degree + 1;
+    poly[roots_len] = FR_ONE;
+    *poly_len = roots_len + 1;
 
-out:
-    return ret;
+    return C_KZG_OK;
 }
 
-/**
- * Calculate the minimal polynomial that evaluates to zero for the powers of
- * roots of unity that correspond to missing indices.
+/* Computes the minimal polynomial that evaluates to zero at equally
+ * spaced chosen roots of unity in the domain of size `FIELD_ELEMENTS_PER_BLOB`.
  *
- * This is done simply by multiplying together `(x - r^i)` for all the `i` that
- * are missing indices, using a combination of direct multiplication
- * (#do_zero_poly_mul_partial) and iterated multiplication via convolution
- * (#reduce_partials).
+ * The roots of unity are chosen based on the missing cell indices. If the i'th
```

Comment:
I think the doc comment (and comments in the function itself) here is already pretty good. 
One thing I would add is where to get the `i'th root of unity` from, as this heavily depends on the order of the roots. 

If I understand correctly, it would be `s->expanded_roots_of_unity[i]`?

Reply:
Yep I can modify that to be more explicit, and yep it is `expanded_roots_of_unity` (non-bit-reversed) -- note that the code has been modularized so that all of the vanishing polynomial calculations/functions do not depend on any bit reversal permutation

Reply:
I've modified it to include `expanded_roots_of_unity`, though feel free to push a commit to modify it further